### DISCUSSION TITLE
Memory limit set and exceeded

### DISCUSSION THREAD

**Discussion by AndyHFW** [ImageMagick/ImageMagick#6819] [Help]
### ImageMagick version

6.9.10.97

### Operating system, version and so on

Amazon Linux 2

### Description

Hi all,

I'm working on an service that uses ImageMagick running on Amazon Linux 2 in a k8s cluster. This cluster has cpu/memory limits set that will kill the service if it exceeds memory limits.

The amount of memory available to ImageMagick is ~2.9 GB and the resource limits are set as:
```
Resource limits:
  Width: 50KP
  Height: 50KP
  List length: 128B
  Area: 53.329GP
  Memory: 2.14844GiB
  Map: 49.6666GiB
  Disk: unlimited
  File: 786432
  Thread: 16
  Throttle: 0
  Time: 3440
```

The issue I'm seeing is that setting this memory limit doesn't seem to prevent the service from being killed.

For reference, the image I'm working with is 14552 x 17585 (~1 GB .tif file) and the convert command is:
`convert tif:/file.tif[0] -auto-orient -resize "1366x9999>" -quality 10 -colorspace sRGB -units PixelsPerInch -density 326 -transparent white png:/converted.png`

I ran the above command with `-debug resource` and it looks like ImageMagick acquires 2.38 GiB of memory and then proceeds to pass it to disk/memory map once it realizes it's over the memory limit:
```
Resource convert[122]: resource.c/AcquireMagickResource/434/Resource
  Memory: 2.38323GiB/0B/2.14844GiB
2023-10-26T20:33:18+00:00 0:01.363 0.680u 6.9.10 Resource convert[122]: resource.c/AcquireMagickResource/434/Resource
  Disk: 2.38323GiB/2.38323GiB/8EiB
2023-10-26T20:33:18+00:00 0:01.363 0.680u 6.9.10 Resource convert[122]: resource.c/AcquireUniqueFileResource/627/Resource
  ...
2023-10-26T20:33:18+00:00 0:01.363 0.680u 6.9.10 Resource convert[122]: resource.c/AcquireUniqueFileResource/683/Resource
  Acquire /tmp/magick-122OwJUtKUdRo1X
2023-10-26T20:33:18+00:00 0:01.363 0.680u 6.9.10 Resource convert[122]: resource.c/AcquireMagickResource/434/Resource
  File: 1B/1B/786432B
2023-10-26T20:33:18+00:00 0:01.365 0.690u 6.9.10 Resource convert[122]: resource.c/AcquireMagickResource/434/Resource
  Map: 2.38323GiB/3.35049GiB/49.6666GiB
2023-10-26T20:33:18+00:00 0:01.365 0.690u 6.9.10 Resource convert[122]: resource.c/RelinquishMagickResource/1256/Resource
  Map: 2.55897GB/990.474MiB/49.6666GiB
  ```
If we assume each pixel requires 8 bytes of memory (Q16), a 14552 x 17585 image would require ~2047 MB, which is less than what was actually acquired above. Additionally, while this command was running, I saw memory use hover between 4.5 GB to 5.5 GB (idle being ~.6 GB).

Ultimately, the questions are;

1. Is there a way to strictly prevent ImageMagick from exceeding the memory limit?
2. Given the memory limit of 2.15 GiB (and subsequent transfer to memory map), why would memory use spike to >4 GB?

**Comment by AndyHFW** [comment]
Barring a different way to manage the memory, two possible avenues I'm looking at are:

1. Increase the allocated memory (to a limit)
2. Use some sort of tiling/streaming of large images as described in https://www.imagemagick.org/Usage/files/#massive

I'd like to better understand the memory limiting/usage before deciding though.

**Comment by urban-warrior** [comment]
Memory limits manage the pixel cache.  You can manage some heap allocations with the system:max-memory-request policy.  ImageMagick does not control other memory allocations and does not control memory allocated by the delegate libraries, e.g., PNG, JPEG, etc.  You might benefit from reading https://imagemagick.org/script/architecture.php#tera-pixel.  You could force the pixel cache of images to disk, freeing up heap memory for other uses.

**Comment by AndyHFW** [comment]
For future reference, I did some testing on Ubuntu 20.04. It's very anecdotal and not particularly scientific, but I'll share it in case it's helpful to anyone looking in later.

- Memory usage was observed using `sudo htop`
- TestA's were using .tif files that had failed in the initial testing
- TestB's were generated using python3-pil, using white pixels (RGB for most files, monochrome for B-3)
- `Estimated size of pixel cache` was calculated using pixels*8 (for Q16 ImageMagick)
- `Actual memory size` was recorded from `debug -resource`
- Memory, map, and disk limits were set in `policy.xml`
- "Peak 1" was generally before the command split into multiple processes
- "Peak 2" was generally right at the end
- Each test was only run a single time, meaning results may not be reliable

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/AndyWang/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/AndyWang/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



File | Filesize (MB) | Width | Height | Pixels | Estimated size of pixel cache (MiB) | Actual memory size (MiB) | Memory limit (MiB) | Map limit | Disk limit | Max memory consumption peak 1 (MiB) | Max memory consumption peak 2 (MiB) |  
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
TestA-1 (1 GB) | 990 | 14552 | 17585 | 255896920 | 1952.338562 | 2272 | 100 | 49 GiB | unlimited | 2949 | 2582 |  
TestA-1 (1 GB) | 990 | 14552 | 17585 | 255896920 | 1952.338562 | 2272 | 1000 | 49 GiB | unlimited | 2363 | 2595 |  
TestA-1 (1 GB) | 990 | 14552 | 17585 | 255896920 | 1952.338562 | 2272 | 10000 | 49 GiB | unlimited | 4751 | 2593 |  
TestA-1 (1 GB) | 990 | 14552 | 17585 | 255896920 | 1952.338562 | 2272 | 100 | 0 | unlimited | 2074 | 2582 |  
TestA-1 (1 GB) | 990 | 14552 | 17585 | 255896920 | 1952.338562 | 2272 | 1000 | 0 | unlimited | 2293 | 2595 |  
TestA-1 (1 GB) | 990 | 14552 | 17585 | 255896920 | 1952.338562 | 2272 | 10000 | 0 | unlimited | 4611 | 2591 |  
TestB-1 (4 GB) | 4264 | 37700 | 37700 | 1421290000 | 10843.58215 | 10099 | 100 | 49 GiB | unlimited | 4068 | 113 | the lack of peak 2 here might be because generated test file is all white
TestB-1 (4 GB) | 4264 | 37700 | 37700 | 1421290000 | 10843.58215 | 10099 | 1000 | 49 GiB | unlimited | 4048 | 503 |  
TestB-1 (4 GB) | 4264 | 37700 | 37700 | 1421290000 | 10843.58215 | 10099 | 10000 | 49 GiB | unlimited | 4043 | 502 |  
TestB-1 (4 GB) | 4264 | 37700 | 37700 | 1421290000 | 10843.58215 | 10099 | 15000 | 49 GiB | unlimited | 13400 | 10500 |  
TestA-2 (1.28 GB) | 1280 | 13956 | 8400 | 117230400 | 894.3969727 | 894 | 100 | 49 GiB | unlimited | 2625 | 2490 | At one point it peaked to 3222 MiB memory (right after the normal peak 1)
TestA-2 (1.28 GB) | 1280 | 13956 | 8400 | 117230400 | 894.3969727 | 894 | 1000 | 49 GiB | unlimited | 2495 | 2523 | 3498
TestA-2 (1.28 GB) | 1280 | 13956 | 8400 | 117230400 | 894.3969727 | 894 | 10000 | 49 GiB | unlimited | 1966 | 2526 | 3491
TestB-2 (3 GB) | 2995 | 31600 | 31600 | 998560000 | 7618.408203 | 7616 | 100 | 49 GiB | unlimited | 2788 | <100 |  
TestB-2 (3 GB) | 2995 | 31600 | 31600 | 998560000 | 7618.408203 | 7616 | 1000 | 49 GiB | unlimited | 2842 | 413 |  
TestB-2 (3 GB) | 2995 | 31600 | 31600 | 998560000 | 7618.408203 | 7616 | 10000 | 49 GiB | unlimited | 8496 | 8006 |  
TestB-3 (monochrome) | 1250 | 100000 | 100000 | 10000000000 | 76293.94531 | 76293.945 | 10000 | 49 GiB | unlimited | 659 | 1187 |  



</body>

</html>

Ultimately, we're planning on bumping our memory limit up (and making the ImageMagick limits more configurable on the fly) before doing more testing.
